hsm/manager_hsm.go

```diff
@@ -0,0 +1,261 @@
+package hsm
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rsa"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+
+	"github.com/ory/fosite"
+	"github.com/ory/hydra/jwk"
+
+	"github.com/miekg/pkcs11"
+
+	"github.com/ory/hydra/x"
+
+	"github.com/ThalesIgnite/crypto11"
+	"github.com/pborman/uuid"
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/cryptosigner"
+)
+
+type KeyManager struct {
+	Context
+}
+
+var _ jwk.Manager = &KeyManager{}
+
+var ErrPreGeneratedKeys = &fosite.RFC6749Error{
+	CodeField:        http.StatusBadRequest,
+	ErrorField:       http.StatusText(http.StatusBadRequest),
+	DescriptionField: "Cannot add/update pre generated keys on Hardware Security Module.",
+}
+
+func NewKeyManager(hsm *Context) *KeyManager {
+	return &KeyManager{
+		*hsm,
+	}
+}
+
+func (m *KeyManager) GenerateKeySet(_ context.Context, set, kid, alg, use string) (*jose.JSONWebKeySet, error) {
+	if len(kid) == 0 {
+		kid = uuid.New()
+	}
+
+	privateAttrSet, publicAttrSet, err := getKeyPairAttributes(kid, set, use)
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case alg == "RS256":
+		key, err := m.GenerateRSAKeyPairWithAttributes(publicAttrSet, privateAttrSet, 4096)
+		if err != nil {
+			return nil, err
+		}
+		return createKeySet(key, kid, alg, use)
+	case alg == "ES256":
+		key, err := m.GenerateECDSAKeyPairWithAttributes(publicAttrSet, privateAttrSet, elliptic.P256())
+		if err != nil {
+			return nil, err
+		}
+		return createKeySet(key, kid, alg, use)
+	case alg == "ES512":
+		key, err := m.GenerateECDSAKeyPairWithAttributes(publicAttrSet, privateAttrSet, elliptic.P521())
+		if err != nil {
+			return nil, err
+		}
+		return createKeySet(key, kid, alg, use)
+	// TODO: HS256, HS512. Makes sense only if shared HSM is used between hydra and authenticating application?
```

Comment:
HS256, HS512 on HSM makes sense only when HSM is shared between application. Can this be a use case? Also when I look fosite/jwt then there is only RS256JWTStrategy/ES256JWTStrategy so how are HS256, HS512 used at all?